File: frontend/src/shared/resource-types.js
/**
 * The three kinds of thing a project spends money on, and their names on screen.
 *
 *   material      — bought by quantity: kilograms, cubic metres, pieces.
 *   work          — bought by the hour: crews and machines alike.
 *   general_cost  — a lump sum with no quantity.
 *
 * Until 2026-09-26 the service split `work` into `labor` and `equipment`. No figure ever
 * depended on which, and MS Project — where every resource comes from — calls both WORK
 * without saying. So the service now has three kinds (migration 0038), and this module is
 * the frontend's single statement of them. Every label map that used to spell the four
 * out reads from here.
 *
 * THE OLD NAMES STILL ARRIVE. Rows stored before 0038 keep `labor` or `equipment` in the
 * database, and issued report snapshots are frozen with those words; the service
 * canonicalises what it reads, but a snapshot payload comes back as it was written. So a
 * type met anywhere is passed through `canonicalResourceType` first, and the old names
 * are `work` here exactly as they are on the service.
 */

export const RESOURCE_TYPES = Object.freeze(["material", "work", "general_cost"]);

/** What `work` used to be called. Read as `work`; never written again. */
export const LEGACY_WORK_TYPES = Object.freeze(["labor", "equipment"]);

/** `labor` and `equipment` are `work`; anything else comes back as it came. */
export function canonicalResourceType(value) {
  return LEGACY_WORK_TYPES.includes(value) ? "work" : value;
}

/**
 * The names. Two registers, because a form field and a chart axis do not want the same
 * length: the long one names the thing, the short one fits beside a bar.
 */
export const RESOURCE_TYPE_LABELS = Object.freeze({
  material: "مصالح",
  work: "نیرو و تجهیزات",
  general_cost: "هزینه‌های عمومی پروژه",
});

export const RESOURCE_TYPE_SHORT_LABELS = Object.freeze({
  material: "مصالح",
  work: "نیرو و تجهیزات",
  general_cost: "هزینه عمومی",
});

/** The label for any type the service has ever sent, or null for one it never has. */
export function resourceTypeLabel(value, labels = RESOURCE_TYPE_LABELS) {
  return labels[canonicalResourceType(value)] ?? null;
}

/**
 * Per-type rows folded onto the three kinds: two `labor` and `equipment` rows from an old
 * snapshot become one `work` row whose money fields are summed exactly, as strings of
 * integer rials. `fields` names the keys to sum; every other key is taken from the first
 * row of the group.
 */
export function foldLegacyTypes(rows, fields) {
  const byType = new Map();
  for (const row of rows ?? []) {
    const type = canonicalResourceType(row?.resourceType);
    const existing = byType.get(type);
    if (!existing) {
      byType.set(type, { ...row, resourceType: type });
      continue;
    }
    for (const field of fields) {
      const a = existing[field];
      const b = row[field];
      if (a == null && b == null) continue;
      existing[field] = String(BigInt(a ?? "0") + BigInt(b ?? "0"));
    }
  }
  return [...byType.values()];
}

File: backend/app/finance/domain/resource_types.py
# -*- coding: utf-8 -*-
"""The three kinds of thing a project spends money on, stated once.

    material      -- bought by quantity: kilograms, cubic metres, pieces.
    work          -- bought by the hour: crews and machines alike.
    general_cost  -- a lump sum with no quantity at all.

WHY THREE AND NOT FOUR

Until 2026-09-26 Finance separated `work` into `labor` and `equipment`. Nothing in the
numbers ever depended on which: both are hours, both are priced per hour, both are settled
against invoices by amount over the rate of the invoice's day. The one place the split
mattered was the import, and there it was a cost: MS Project calls both WORK and does not
say which, so every crew and every machine in a schedule waited for a person to decide --
through an endpoint that existed only on the development host. On the product, they waited
forever. Three kinds is what the file states, what the arithmetic needs, and what the
project owner chose: «تجهیزات و نیروی انسانی باید جزو یک دسته بندی باشند».

THE OLD NAMES

Rows written before 0038 still say `labor` or `equipment`, and issued report snapshots are
frozen with those words. They are not rewritten -- the migration widens the CHECK and
touches no row -- so every reader that meets a stored type passes it through
`canonical_resource_type` first. That function is the whole compatibility story: one place,
and a stored `equipment` is `work` everywhere else.
"""

MATERIAL = "material"
WORK = "work"
GENERAL_COST = "general_cost"

#: The kinds, in the order every breakdown lists them.
RESOURCE_TYPES = (MATERIAL, WORK, GENERAL_COST)

#: What `work` used to be called. Accepted when READ, never written again.
LEGACY_WORK_TYPES = ("labor", "equipment")

#: Everything a stored `resource_type` may hold, for a CHECK or a validator that must
#: still admit the rows that exist.
STORED_RESOURCE_TYPES = RESOURCE_TYPES + LEGACY_WORK_TYPES


def canonical_resource_type(value):
    """`labor` and `equipment` are `work`; anything else is returned as it came.

    `None` stays `None`: an unclassified resource is not silently a kind of work.
    """
    if value in LEGACY_WORK_TYPES:
        return WORK
    return value
